dbms/src/DataTypes/DataTypeString.cpp

```diff
@@ -310,4 +310,181 @@ void registerDataTypeString(DataTypeFactory & factory)
     factory.registerSimpleDataType("LONGBLOB", creator, DataTypeFactory::CaseInsensitive);
 }
 
+namespace
+{
+
+using Offset = ColumnString::Offset;
+
+// Returns <offsets_stream, chars_stream>.
+template <typename B, typename G>
+std::pair<B *, B *> getStream(const G & getter, IDataType::SubstreamPath & path)
+{
+    auto * chars_stream = getter(path);
+    path.emplace_back(IDataType::Substream::StringSizes);
+    auto * offsets_stream = getter(path);
+    return {offsets_stream, chars_stream};
+}
+
+PaddedPODArray<Offset> offsetToStrSize(
+    const ColumnString::Offsets & chars_offsets,
+    const size_t begin,
+    const size_t end)
+{
+    assert(!chars_offsets.empty());
+    PaddedPODArray<Offset> str_sizes(end - begin);
+    size_t i = 0;
+    if (begin == 0)
+    {
+        str_sizes[0] = chars_offsets[0];
+        ++i;
+    }
+    assert(begin + i > 0);
+    // clang-format off
+    #pragma clang loop vectorize(enable)
+    // clang-format on
+    for (; i < str_sizes.size(); ++i)
+    {
+        str_sizes[i] = chars_offsets[begin + i] - chars_offsets[begin + i - 1];
+    }
```

Comment:
the class `PODArrayBase` ensure chars_offsets[-1] is well defined as 0. I think we can simplify the code and get a better vectorized code.

https://github.com/pingcap/tiflash/blob/0c15f1f2acf78938b8aa7c003d65ab2fef9ca315/dbms/src/Common/PODArray.h#L68-L70